apiserver/remotefirewaller/ingressaddresswatcher.go

```diff
@@ -197,10 +250,109 @@ func (w *IngressAddressWatcher) mergeAddresses(addresses set.Strings, c params.R
 			}
 		}
 		if !inUse {
-			result.Remove(address)
+			changed = true
 		}
 	}
-	return result, nil
+	return changed, nil
+}
+
+func (w *IngressAddressWatcher) trackUnit(unit Unit) error {
+	machine, err := w.assignedMachine(unit)
+	if errors.IsNotAssigned(err) {
+		logger.Errorf("unit %q entered scope without a machine assigned - addresses will not be tracked", unit)
+		return nil
+	}
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	w.unitToMachine[unit.Name()] = machine.Id()
+	mData, ok := w.machines[machine.Id()]
+	if ok {
+		// We're already watching the machine, just add this unit.
+		mData.units.Add(unit.Name())
+		return nil
+	}
+
+	addressWatcher, err := newMachineAddressWatcher(machine, w.addressChanges)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	w.machines[machine.Id()] = &machineData{
+		units:   set.NewStrings(unit.Name()),
+		watcher: addressWatcher,
+	}
+	err = w.catacomb.Add(addressWatcher)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (w *IngressAddressWatcher) untrackUnit(unitName string) error {
+	machineId, ok := w.unitToMachine[unitName]
+	if !ok {
+		logger.Errorf("missing machine id for unit %q", unitName)
+		return nil
+	}
+	mData, ok := w.machines[machineId]
+	if !ok {
+		// This shouldn't happen.
+		return errors.Errorf("missing machine data for machine %q (hosting unit %q)", machineId, unitName)
+	}
+	mData.units.Remove(unitName)
+	if mData.units.Size() > 0 {
+		// No need to stop the watcher - there are still units on the
+		// machine.
+		return nil
+	}
+
+	err := worker.Stop(mData.watcher)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	delete(w.machines, machineId)
+	delete(w.unitToMachine, unitName)
+	return nil
+}
+
+func (w *IngressAddressWatcher) assignedMachine(unit Unit) (Machine, error) {
+	machineId, err := unit.AssignedMachineId()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	machine, err := w.backend.Machine(machineId)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return machine, nil
+}
+
+func (w *IngressAddressWatcher) machineAddressChanged(machineId string) (bool, error) {
```

Comment:
The method name implies this is a read only operation.
But "known" is updated.
Can we call it processMachineAddresses() or something.
The return values could be named to make it clear what is happening eg (changed bool, _ error)